design/ceph-config-updates.md

```diff
@@ -0,0 +1,235 @@
+# Updating Rook's Ceph configuration strategy (Jan 2019)
+**Targeted for v1.0-v1.1**
+
+## Background
+
+Starting with Ceph Mimic, Ceph is able to store the vast majority of config options for all daemons
+in the Ceph mons' key-value store. This ability to centrally manage 99% of its configuration
+is Ceph's preferred way of managing config. This conveniently allows Ceph to fit better into the
+containerized application space than before.
+
+However, for it's own backwards compatibility, Ceph options set in the config file or via command
+line flags will override the centrally-configured settings. To make the most of this functionality
+within Ceph, it will be necessary to limit the configuration options specified by Rook in either
+config files or on the command line to a minimum.
+
+
+## Proposed new design
+
+The end goal of this work will be to allow Ceph to centrally manage its own configuration as much as
+possible. Or in other terms, Rook will specify the barest minimum of configuration options in the
+config file or on the command line with a priority on cleaning the config file.
+
+Rook currently offers the option of a config override, and this functionality can be kept for
+backwards compatibility for existing users; however, we should note that it is advisable not to do
+this and instead configure Ceph's centrally-managed options.
+
+### Minimal config file
+
+#### Required minimum
+
+All Ceph options in the config file can be set via the command line, so it is tempting to remove the
+need for having a `ceph.conf` in containers at all. In practice, we do not know how to make this an
+actual possibility.
+
+The troublesome situation regards Ceph daemons' ability to discover mons. The daemons must have
+`mon host` specified to find mons which are using non-default ports. It is further undesirable to
+specify this via the `--mon-host` flag because if this parameter is specified statically as a
+container argument and the initial mons fail and are replaced by new mons, the daemon will not know
+how to contact the mons if it is restarted.
+
+The proposed solution to this problem and the only config option which have thus far been deemed
+*required* is the `mon host` global option which will be stored in a Kubernetes ConfigMap so that
+containers always have the most updated version of this information. The single `ceph.conf`
+ConfigMap will be mounted to all containers at `/etc/ceph/ceph.conf`. There is a possibility the
+`mon initial members` global option may also be necessary, as it goes hand-in-hand with `mon host`.
+
+With this one -- or these two -- options set via the config file, all daemons should *require* no
```

Comment:
A configmap value can also be passed as an environment variable to the pod. Could we use something like `--mon-host=$(ROOK_MON_ENDPOINTS)`?